### PR TITLE
frontend: mitigate blocking the main thread

### PR DESCRIPTION
**Pull request by yyyyaaa**
#### What's this PR do?

- [x] Add some notes about prevent blocking the main thread for frontend perf

### CODE REVIEW

File: Engineering/Frontend/Mitigate blocking the main thread.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+---
+tags: javascript, frontend performance, engineering/frontend
```
Comment (zlatanpham) [inline Engineering/Frontend/Mitigate blocking the main thread.md:2]: A tag doesn't work with space

```suggestion
tags: javascript, frontend/performance, engineering/frontend
```

File: Engineering/Frontend/Mitigate blocking the main thread.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+With Web Worker we can offload a bunch of processing to another thread and it will not impact the performance of the main thread. Sounds perfect, but it comes with a cost: `serialization`.
+
+Because of the heavy sandboxing of JS environment, using web workers only works well if we have small to medium inputs and outputs. Because all of the data transported between the main thread and worker thread is going to be serialized, which blocks the main thread while that happens (unless you are using binary formats like Shared Array Buffer or Typed Arrays, which comes with other caveats that is outside of the scope of this note).
+
+If `serialization` is not a problem for your app then Web Worker will probably do a great job to enhance your app's performance and UX.
+
+For real time apps that require continuously recalculation of data when new data comes, you can implement a game-loop-like updater function inside the web worker code to periodically emit data to main thread once per an interval time limit. That pattern works well because it allows you to control how often main thread should receive new data and UI should rerender.
+
+### Coroutines with generators
```
Comment (zlatanpham) [inline Engineering/Frontend/Mitigate blocking the main thread.md:24]: Looking forward to a demo 😎 (if you plan to)

**Review by zlatanpham — APPROVED** [review]
Great first contribution @yyyyaaa 🎉

File: Engineering/Frontend/Mitigate blocking the main thread.md (hunk 2 above)
Comment (yyyyaaa) [reply Engineering/Frontend/Mitigate blocking the main thread.md:24]: I can copy the worker code from nghenhan and make a codesandbox lol, will add tmr

Comment (zlatanpham) [reply Engineering/Frontend/Mitigate blocking the main thread.md:24]: If you can do a demo for it this Friday, it'd be great ⭐

**Pull request merged**